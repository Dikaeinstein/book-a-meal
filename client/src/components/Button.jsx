import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ value, className }) => (
  <button className={className}>
    {value}
  </button>
);

Button.propTypes = {
  value: PropTypes.string.isRequired,
  className: PropTypes.string.isRequired,
};

export default Button;
