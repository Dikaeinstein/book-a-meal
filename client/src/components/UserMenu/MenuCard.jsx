import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Preloader } from 'react-materialize';
import { fetchMenu } from '../../actions/menuActions';
import ConnectedMealList from './MealList';
import Loading from '../util/Loading';
import Paginate from '../util/Paginate';
import {
  getNextPageUrl,
  getPreviousPageUrl, getCurrentPageUrl,
} from '../../reducers/paginationReducer';

export class MenuCard extends Component {
  constructor(props) {
    super(props);
    this.handlePageChange = this.handlePageChange.bind(this);
  }

  componentDidMount() {
    this.handlePageChange(this.props.currentUrl);
  }

  handlePageChange(url) {
    this.props.fetchMenu(url);
  }

  render() {
    const { isFetching, link } = this.props;
    if (isFetching) {
      return (
        <div className="loader-container">
          <Loading text="fetching menu...">
            <Preloader flashing size="medium" />
          </Loading>
        </div>
      );
    }

    return (
      <section className="main" id="menu">
        <h4 className="text-center">
          Menu for Today {(new Date()).toDateString()}
        </h4>
        <ConnectedMealList link={link} />
        <Paginate
          onPageChange={this.handlePageChange}
          nextUrl={this.props.nextUrl}
          previousUrl={this.props.previousUrl}
          style={{ marginTop: '0' }}
        />
      </section>
    );
  }
}

MenuCard.defaultProps = {
  nextUrl: '', previousUrl: '',
};

MenuCard.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  fetchMenu: PropTypes.func.isRequired,
  link: PropTypes.string.isRequired,
  currentUrl: PropTypes.string.isRequired,
  nextUrl: PropTypes.string,
  previousUrl: PropTypes.string,
};

const mapStateToProps = state => ({
  isFetching: state.menu.isFetching,
  nextUrl: getNextPageUrl(state.pagination.menu),
  currentUrl: getCurrentPageUrl(state.pagination.menu),
  previousUrl: getPreviousPageUrl(state.pagination.menu),
});

export default connect(mapStateToProps, { fetchMenu })(MenuCard);
