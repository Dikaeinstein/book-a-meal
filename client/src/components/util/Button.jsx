import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ value, ...otherProps }) => (
  <button {...otherProps}>
    {value}
  </button>
);

Button.propTypes = {
  value: PropTypes.string.isRequired,
  className: PropTypes.string.isRequired,
};

export default Button;
