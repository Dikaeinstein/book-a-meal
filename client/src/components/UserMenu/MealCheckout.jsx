import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import history from '../../helpers/history';
import Button from '../util/Button';
import { checkoutOrder } from '../../actions/orderActions';

class ConnectedMealCheckout extends Component {
  constructor(props) {
    super(props);
    this.state = {
      total: parseFloat(props.meal.price),
      quantity: 1,
      amount: parseFloat(props.meal.price),
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleCheckout = this.handleCheckout.bind(this);
  }

  handleChange(event) {
    event.preventDefault();
    const value = event.target.value;
    if (value && value >= 1) {
      const newQuantity = parseInt(value, 10);
      this.setState({
        quantity: newQuantity,
        total: newQuantity * this.state.amount,
      });
    }
  }

  handleCheckout() {
    this.props.closeModal();
    this.props.checkoutOrder({
      mealId: this.props.meal.id,
      name: this.props.meal.name,
      quantity: this.state.quantity,
      total: this.state.total,
      amount: this.props.meal.price,
    });
    history.push('/order-confirmation');
  }

  render() {
    return (
      <div id="checkout" className="text-center">
        <h3>Checkout Order</h3>
        <div>
          <table>
            <thead className="text-center">
              <tr>
                <th>
                  Meal Name
                </th>
                <th>
                  Quantity (Plate)
                </th>
                <th>
                  Total
                </th>
              </tr>
            </thead>
            <tbody>
              <tr className="text-center">
                <td>
                  {this.props.meal.name}
                </td>
                <td>
                  <input
                    type="number"
                    name="quantity"
                    id="quantity"
                    value={this.state.quantity}
                    onChange={this.handleChange}
                    min={1}
                    max={Number.MAX_SAFE_INTEGER}
                    style={{ width: '50%' }}
                  />
                </td>
                <td>
                  &#x20a6; {this.state.total}
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <Button
          className="btn btn-default"
          value="Checkout"
          style={{ margin: '1rem auto' }}
          onClick={this.handleCheckout}
        />
      </div>
    );
  }
}

ConnectedMealCheckout.propTypes = {
  meal: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ])).isRequired,
  closeModal: PropTypes.func.isRequired,
  checkoutOrder: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  checkoutOrder(order) { dispatch(checkoutOrder(order)); },
});

const MealCheckout = connect(null, mapDispatchToProps)(ConnectedMealCheckout);

export default MealCheckout;
