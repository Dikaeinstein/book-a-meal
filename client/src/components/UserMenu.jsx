import React from 'react';
import MenuCard from './MenuCard';

const UserMenu = () => <MenuCard link="/order-confirmation" />;

export default UserMenu;
