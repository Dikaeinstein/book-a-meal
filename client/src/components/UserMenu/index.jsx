import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MenuCard from './MenuCard';
import Footer from '../util/Footer';
import errorHandler from '../util/errorHandler';

const ConnectedUserMenu = ({ fetchMenuError }) => {
  const MenuCardWithErrorHandling =
    errorHandler(MenuCard, 'Error fetching menu');

  return (
    <div>
      <div className="bg-dark-light landing-main">
        <main style={{ minHeight: 'calc(100vh - 151px)' }}>
          <MenuCardWithErrorHandling
            link="/order-confirmation"
            error={fetchMenuError}
          />
        </main>
      </div>
      <Footer />
    </div>
  );
};

ConnectedUserMenu.propTypes = {
  /* eslint react/require-default-props: 0 */
  fetchMenuError: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.objectOf(PropTypes.string),
  ]),
};

const mapStateToProps = state => ({
  fetchMenuError: state.menus.fetchError,
});

const UserMenu = connect(mapStateToProps)(ConnectedUserMenu);

export default UserMenu;
