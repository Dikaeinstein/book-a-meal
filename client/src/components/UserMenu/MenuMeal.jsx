import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import Modal from 'react-modal';
import MealDetail from '../Meals/MealDetail';
import MealCheckout from './MealCheckout';
import SigninSignupButtons from './SigninSignupButtons';
import Button from '../util/Button';


class MenuMeal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isMealDetailOpen: false,
      isMealChoiceOpen: false,
      meal: props.meal,
    };
    this.handleOpenMealDetailModal =
      this.handleOpenMealDetailModal.bind(this);
    this.handleCloseMealDetailModal =
      this.handleCloseMealDetailModal.bind(this);
    this.handleOpenMealChoiceModal =
      this.handleOpenMealChoiceModal.bind(this);
    this.handleCloseMealChoiceModal =
      this.handleCloseMealChoiceModal.bind(this);
  }

  handleOpenMealDetailModal() {
    this.setState({ isMealDetailOpen: true });
  }

  handleCloseMealDetailModal() {
    this.setState({ isMealDetailOpen: false });
  }

  handleOpenMealChoiceModal() {
    this.setState({ isMealChoiceOpen: true });
  }

  handleCloseMealChoiceModal() {
    this.setState({ isMealChoiceOpen: false });
  }

  render() {
    const modalStyle = {
      overlay: {
        backgroundColor: 'rgba(0, 0, 0, 0.25)',
      },
      content: {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        marginRight: '-50%',
        transform: 'translate(-50%, -50%)',
        minWidth: '20rem',
        width: '70%',
      },
    };
    const mealChoiceModalStyle = {
      ...modalStyle,
      content: { ...modalStyle.content, width: '50%' },
    };
    return (
      <div key={this.state.meal.id} className="card user-menu-item">
        <div
          className="overlay-container"
          onClick={this.handleOpenMealDetailModal}
          onKeyPress={this.handleOpenMealDetailModal}
          role="button"
          tabIndex="0"
        >
          <img
            src={this.state.meal.imageUrl}
            alt={this.state.meal.name}
          />
          <div className="overlay">
            <div className="text">
              <p>View description</p>
            </div>
          </div>
        </div>
        <Modal
          isOpen={this.state.isMealDetailOpen}
          contentLabel="Meal Detail"
          style={modalStyle}
          closeTimeoutMS={150}
        >
          <Button
            value="&times;"
            onClick={this.handleCloseMealDetailModal}
            className="close"
          />
          <MealDetail
            closeMealDetail={this.handleCloseMealDetailModal}
            meal={this.state.meal}
          />
        </Modal>
        <p className="text-black">{this.state.meal.name}</p>
        <p className="text-black">&#x20a6;{this.state.meal.price}</p>
        <Modal
          isOpen={this.state.isMealChoiceOpen}
          contentLabel="Meal Choice"
          style={mealChoiceModalStyle}
          closeTimeoutMS={150}
        >
          <Button
            value="&times;"
            onClick={this.handleCloseMealChoiceModal}
            className="close"
          />
          {this.props.loggedIn
            ?
              <MealCheckout
                meal={this.state.meal}
                closeModal={this.handleCloseMealChoiceModal}
              />
            :
              <SigninSignupButtons
                closeModal={this.handleCloseMealChoiceModal}
              />
          }
        </Modal>
        <Button
          className="btn btn-default"
          onClick={this.handleOpenMealChoiceModal}
          value="Order Now"
          style={{ margin: '1rem auto' }}
        />
      </div>
    );
  }
}

MenuMeal.propTypes = {
  meal: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ])).isRequired,
  loggedIn: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  loggedIn: state.user.loggedIn,
});


export default connect(mapStateToProps)(MenuMeal);
