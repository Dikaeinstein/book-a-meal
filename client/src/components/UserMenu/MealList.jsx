import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MenuMeal from './MenuMeal';
import { getMenu } from '../../reducers/menuReducer';


const ConnectedMealList = ({ meals, link }) => {
  const mealList = meals
    .map(meal => <MenuMeal key={meal.id} defaultMeal={meal} link={link} />);

  return (
    <section className="card user-menu text-left">
      {mealList}
    </section>
  );
};

ConnectedMealList.propTypes = {
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  link: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  meals: getMenu(state.menu).meals,
});

const MealList = connect(mapStateToProps)(ConnectedMealList);

export default MealList;
