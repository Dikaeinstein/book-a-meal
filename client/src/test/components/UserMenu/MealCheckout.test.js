import React from 'react';
import { shallow } from 'enzyme';
import { MealCheckout } from '../../../components/UserMenu/MealCheckout';
import Button from '../../../components/util/Button';


describe('<MealCheckout />', () => {
  const closeModal = jest.fn();
  it('should handle order checkout', () => {
    const checkoutOrder = jest.fn();
    const wrapper = shallow(<MealCheckout
      defaultMeal={{ price: 3000 }}
      closeModal={closeModal}
      checkoutOrder={checkoutOrder}
    />);
    const checkout = wrapper.find(Button);
    checkout.simulate('click');
    expect(closeModal).toHaveBeenCalled();
  });
  it('should handle meal quantity change', () => {
    const checkoutOrder = jest.fn();
    const wrapper = shallow(<MealCheckout
      defaultMeal={{ price: 3000 }}
      closeModal={closeModal}
      checkoutOrder={checkoutOrder}
    />);
    expect(wrapper.state().total).toEqual(3000);
    expect(wrapper.state().quantity).toEqual(1);
    wrapper.find('#quantity').simulate('change', {
      preventDefault: jest.fn(),
      target: { value: 2 },
    });
    expect(wrapper.state().total).toEqual(6000);
    expect(wrapper.state().quantity).toEqual(2);
  });
  it('should match snapshot', () => {
    const checkoutOrder = jest.fn();
    const tree = shallow(<MealCheckout
      defaultMeal={{ price: 3000 }}
      closeModal={jest.fn()}
      checkoutOrder={checkoutOrder}
    />);
    expect(tree).toMatchSnapshot();
  });
});
