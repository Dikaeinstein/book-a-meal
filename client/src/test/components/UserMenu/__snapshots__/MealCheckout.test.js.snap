// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MealCheckout /> should match snapshot 1`] = `
<div
  className="text-center"
  id="checkout"
>
  <h5>
    Checkout Order
  </h5>
  <div>
    <table>
      <thead
        className="text-center"
      >
        <tr>
          <th>
            Meal Name
          </th>
          <th>
            Quantity (Plate)
          </th>
          <th>
            Total
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="text-center"
        >
          <td />
          <td>
            <input
              id="quantity"
              max={9007199254740991}
              min={1}
              name="quantity"
              onChange={[Function]}
              style={
                Object {
                  "width": "50%",
                }
              }
              type="number"
              value={1}
            />
          </td>
          <td>
            â‚¦ 
            3000
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <Button
    className="btn btn-default"
    onClick={[Function]}
    style={
      Object {
        "margin": "1rem auto",
      }
    }
    value="Checkout"
  />
</div>
`;
